homeassistant/components/vacuum/ecovacs.py

```diff
@@ -0,0 +1,195 @@
+        """Return true if vacuum is currently charging."""
+        return self.device.is_charging
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag vacuum cleaner robot features that are supported."""
+        return SUPPORT_ECOVACS
+
+    @property
+    def status(self):
+        """Return the status of the vacuum cleaner."""
+        return self.device.vacuum_status
+
+    def return_to_base(self, **kwargs):
+        """Set the vacuum cleaner to return to the dock."""
+        from sucks import Charge
+        self.device.run(Charge())
+
+    @property
+    def battery_icon(self):
+        """Return the battery icon for the vacuum cleaner."""
+        return icon_for_battery_level(
+            battery_level=self.battery_level, charging=self.is_charging)
+
+    @property
+    def battery_level(self):
+        """Return the battery level of the vacuum cleaner."""
+        if self.device.battery_status is not None:
+            return self.device.battery_status * 100
+
+        return super().battery_level
+
+    @property
+    def fan_speed(self):
+        """Return the fan speed of the vacuum cleaner."""
+        return self.device.fan_speed
+
+    @property
+    def fan_speed_list(self):
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        from sucks import FAN_SPEED_NORMAL, FAN_SPEED_HIGH
+        return [FAN_SPEED_NORMAL, FAN_SPEED_HIGH]
+
+    def turn_on(self, **kwargs):
+        """Turn the vacuum on and start cleaning."""
+        from sucks import Clean
+        self.device.run(Clean())
+
+    def turn_off(self, **kwargs):
+        """Turn the vacuum off stopping the cleaning and returning home."""
+        self.return_to_base()
+
+    def stop(self, **kwargs):
+        """Stop the vacuum cleaner."""
+        from sucks import Stop
+        self.device.run(Stop())
+
+    def clean_spot(self, **kwargs):
+        """Perform a spot clean-up."""
+        from sucks import Spot
+        self.device.run(Spot())
+
+    def locate(self, **kwargs):
+        """Locate the vacuum cleaner."""
+        from sucks import PlaySound
+        self.device.run(PlaySound())
+
+    def set_fan_speed(self, fan_speed, **kwargs):
+        """Set fan speed."""
+        if self.is_on:
+            from sucks import Clean
+            self.device.run(Clean(
+                mode=self.device.clean_status, speed=fan_speed))
+
+    def send_command(self, command, params=None, **kwargs):
+        """Send a command to a vacuum cleaner."""
+        from sucks import VacBotCommand
+        self.device.run(VacBotCommand(command, params))
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes of the vacuum cleaner."""
+        data = super().state_attributes
+
+        data[ATTR_ERROR] = self._error
+
+        for key, val in self.device.components.items():
```

Comment:
It's not clear to me what values are added to the state machine. Meaning it can't be documented nor can it be judged if they are appropriate to be stored as state attributes to begin with.

Reply:
Ah, these are attributes representing the lifespan of replaceable components on the vacuum. Today, they are hard coded in the lib to be `main_brush`, `side_brush`, and `filter`. (In Hass the attributes will be `component_main_brush`, etc.) The value is a 0-100 integer representing the percentage of lifespan remaining. I made it dynamic here because those three components may not be present on all vacuum models, so as we get feedback and make `sucks` more comprehensive for all vacuum models, this should appropriately populate based on what the vacuum actually has.

(I plan on documenting all of this in the docs once I implement the config entries)